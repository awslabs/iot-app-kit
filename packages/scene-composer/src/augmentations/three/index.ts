export * from './Anchor';
export * from './Viewpoint';
