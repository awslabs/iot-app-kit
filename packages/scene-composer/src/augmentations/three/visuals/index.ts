export * from './IVisual';
export * from './WidgetVisual';
export * from './WidgetSprite';
