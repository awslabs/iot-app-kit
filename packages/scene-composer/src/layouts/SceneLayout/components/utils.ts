export enum Direction {
  Left = 'Left',
  Right = 'Right',
}
