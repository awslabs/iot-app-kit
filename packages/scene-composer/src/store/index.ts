export * from './StoreOperations';
export * from './Store';
