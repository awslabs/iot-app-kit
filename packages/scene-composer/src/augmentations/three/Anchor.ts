import * as THREE from 'three';

import DebugLogger from '../../logger/DebugLogger';
import { DefaultAnchorStatus, SelectedAnchor } from '../../interfaces';

import { IVisual } from './visuals';

/**
 * The generic anchor object.
 */
export class Anchor extends THREE.Object3D {
  private log = new DebugLogger('Anchor');

  protected visualMap = new Map<string, IVisual>();
  protected _visualState: DefaultAnchorStatus | string = DefaultAnchorStatus.Info;
  protected _isSelected = false;

  public add(visualWithStateName: IVisual): this {
    const stateName = visualWithStateName.name;
    if (!stateName) {
      this.log?.error('A name must be provided for the visual. This is used to determine its appearance.');
      return this;
    }

    if (!visualWithStateName.visual) {
      this.log?.error('A visual must be provided.');
      return this;
    }

    // NOTE: Add the states directly, because event bubbling to the anchor level does not seem to occur on click
    //  and is instead caught by the WidgetVisual
    super.add(visualWithStateName.visual);

    this.visualMap.set(stateName, visualWithStateName);

    // Update the anchor visual with the current visual state
    this.updateVisualState();

    return this;
  }

  public set visualState(visualState: DefaultAnchorStatus | string) {
    this._visualState = visualState;
    this.updateVisualState();
  }

  public get visualState(): DefaultAnchorStatus | string {
    return this._visualState;
  }

  /**
   * Gets the anchor's selected state
   */
  public get isSelected(): boolean {
    return this._isSelected;
  }

  /**
   * Sets the anchor's selected state keying the visibility of the specially named Selected Visual
   * @param value true if the anchor is selected
   */
  public set isSelected(value: boolean) {
    this._isSelected = value;
    this.visualMap.get(SelectedAnchor)?.setVisible(value);
  }

  private updateVisualState() {
    this.visualMap.forEach((visual, name) => {
      visual.setVisible(name === this.visualState);
    });
  }
}
