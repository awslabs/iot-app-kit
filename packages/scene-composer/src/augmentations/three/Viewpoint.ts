import * as THREE from 'three';

import DebugLogger from '../../logger/DebugLogger';

import { WidgetVisual } from './visuals';

export enum ViewpointState {
  Deselected = 'Deselected',
  Selected = 'Selected',
}

/**
 * The generic Viewpoint object.
 */
export class Viewpoint extends THREE.Object3D {
  private log = new DebugLogger('Viewpoint');

  protected visualMap = new Map<string, WidgetVisual>();
  protected _visualState: string = ViewpointState.Deselected;

  constructor() {
    super();
    this.type = 'Viewpoint';
    this.rotateX(THREE.MathUtils.degToRad(90));
  }

  // @ts-ignore
  public add(visualWithStateName: WidgetVisual): this {
    const stateName = visualWithStateName.name;
    if (!stateName) {
      this.log?.error('A name must be provided for the visual. This is used to determine its appearance.');
      return this;
    }

    if (!visualWithStateName.visual) {
      this.log?.error('A visual must be provided.');
      return this;
    }

    // NOTE: Add the states directly, because event bubbling to the anchor level does not seem to occur on click
    //  and is instead caught by the WidgetVisual
    super.add(visualWithStateName.visual);

    this.visualMap.set(stateName, visualWithStateName);

    // Update the visual with the current visual state
    this.updateVisualState();

    return this;
  }

  public set isSelected(value: boolean) {
    this._visualState = value ? ViewpointState.Selected : ViewpointState.Deselected;
    this.updateVisualState();
  }

  public get isSelected(): boolean {
    return this._visualState === ViewpointState.Selected;
  }

  private updateVisualState() {
    this.visualMap.forEach((visual, name) => {
      visual.setVisible(name === this._visualState);
    });
  }
}

export class ViewCursor extends Viewpoint {
  constructor() {
    super();
    this.type = 'ViewCursor';
  }
}
