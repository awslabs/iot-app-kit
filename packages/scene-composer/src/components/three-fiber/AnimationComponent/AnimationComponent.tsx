// eslint-disable-next-line import/default
import React, { Fragment, useEffect, useMemo } from 'react';
import * as THREE from 'three';
import { useFrame } from '@react-three/fiber';

import { useSceneComposerId } from '../../../common/sceneComposerIdContext';
import { ISceneNodeInternal, useEditorState, useViewOptionState } from '../../../store';
import { IAnimationComponent } from '../../../interfaces';

export const toggleIsAnimationPaused = (
  AnimationMixer: THREE.AnimationMixer,
  isPaused: boolean,
  AnimationList: string[],
) => {
  const object = AnimationMixer.getRoot() as THREE.Object3D;
  AnimationList.forEach(function (clip) {
    const animation = THREE.AnimationClip.findByName(object.animations, clip);
    if (AnimationMixer.clipAction(animation)) {
      AnimationMixer.clipAction(animation).paused = isPaused;
    }
  });
};

interface AnimationComponentProps {
  component: IAnimationComponent;
  node: ISceneNodeInternal;
}
export const animationObjectKey = 'Scene';
const AnimationComponent: React.FC<AnimationComponentProps> = ({ component, node }) => {
  const sceneComposerId = useSceneComposerId();
  const isGlobalAnimationPaused = !useViewOptionState(sceneComposerId).componentVisibilities['Animation'];
  const [scene, setScene] = React.useState<THREE.Object3D>();
  const { getObject3DBySceneNodeRef } = useEditorState(sceneComposerId);
  const object = getObject3DBySceneNodeRef(node.ref);
  const currentAnimations = useMemo(() => {
    const currentAnimations: string[] = [];
    if (object) {
      // gltf loader sets this name for the object 3ds it loads
      setScene(object.getObjectByName(animationObjectKey));
      component.currentAnimations?.forEach((animation) => {
        currentAnimations.push(animation);
      });
    }
    return currentAnimations;
  }, [object, component]);

  const AnimationMixer = useMemo(() => {
    let AnimationMixer;
    if (scene) {
      AnimationMixer = new THREE.AnimationMixer(scene);
    }
    return AnimationMixer;
  }, [object, scene, component, node]);

  useEffect(() => {
    if (scene?.animations) {
      currentAnimations.forEach((clip) => {
        const animation = THREE.AnimationClip.findByName(scene.animations, clip);
        AnimationMixer.clipAction(animation)?.play();
      });
      toggleIsAnimationPaused(AnimationMixer, isGlobalAnimationPaused, currentAnimations);
    }
  }, [scene, isGlobalAnimationPaused, currentAnimations, component, AnimationMixer]);
  useFrame((state, delta) => {
    AnimationMixer?.update(delta);
  });
  return <Fragment />;
};

AnimationComponent.displayName = 'AnimationComponent';

export default AnimationComponent;
