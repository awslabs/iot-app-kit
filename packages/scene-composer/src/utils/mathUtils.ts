import * as THREE from 'three';

import { DistanceUnit } from '../interfaces';
import { Color } from '../models/SceneModels';

/**
 * Format bytes as human-readable text.
 *
 * @param bytes Number of bytes.
 * @param si True to use metric (SI) units, aka powers of 1000. False to use
 *           binary (IEC), aka powers of 1024.
 * @param dp Number of decimal places to display.
 *
 * @return Formatted string.
 */
export function humanFileSize(bytes: number, si = false, dp = 2) {
  const thresh = si ? 1000 : 1024;

  if (Math.abs(bytes) < thresh) {
    return bytes + ' B';
  }

  const units = si
    ? ['kB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB']
    : ['KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB', 'ZiB', 'YiB'];
  let u = -1;
  const r = 10 ** dp;

  do {
    bytes /= thresh;
    ++u;
  } while (Math.round(Math.abs(bytes) * r) / r >= thresh && u < units.length - 1);

  return bytes.toFixed(dp) + ' ' + units[u];
}

const PADDINGS = ['0', '00', '000', '0000', '00000'];
/**
 * Convert javascript integer to the css-style hex string representation.
 * For example: 1193046 => 0x123456 => #123456
 *
 * @param dec decimal color representation
 * @returns css-style hex string of the same color
 */
export function decToHexString(dec: number) {
  const str = dec.toString(16);
  const paddingLen = 6 - str.length;
  if (paddingLen > 0) {
    return '#' + PADDINGS[paddingLen - 1] + str;
  }
  return '#' + str;
}

export function colorToHexString(color: Color) {
  if (typeof color === 'string') {
    // assume the string color is css-style hex
    return color;
  }

  return decToHexString(color);
}

export function parseFloatOrDefault(str: string, defaultValue: number) {
  const result = Number.parseFloat(str);
  if (Number.isNaN(result)) {
    return defaultValue;
  }
  return result;
}

export function generateUUID() {
  return THREE.MathUtils.generateUUID();
}

const DISTANCE_TO_METER_MAP: Record<DistanceUnit, number> = {
  millimeters: 0.001,
  centimeters: 0.01,
  decimeters: 0.1,
  meters: 1,
  kilometers: 1000.0,
  inches: 0.0254,
  feet: 0.3048,
  yards: 0.9144,
  miles: 1609.34,
};

/**
 * Return the scale factor to convert `from` unit to `to` unit.
 */
export function getScaleFactor(from: DistanceUnit, to: DistanceUnit) {
  return DISTANCE_TO_METER_MAP[from] / DISTANCE_TO_METER_MAP[to];
}

export const EPSILON = 1e-6;
/**
 * Return whether 2 numbers approximately equal each other.
 * @param a number
 * @param b number
 * @param epsilon Epsilon value to be used in aprroximation comparison. Default is EPSILON
 * @return true is approximately equal
 */
export function approximatelyEquals(a: number, b: number, epsilon = EPSILON): boolean {
  return Math.abs(Math.max(a, b) - Math.min(a, b)) <= epsilon;
}
