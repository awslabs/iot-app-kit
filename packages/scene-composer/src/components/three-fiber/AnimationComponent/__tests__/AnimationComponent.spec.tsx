import * as THREE from 'three';
import React from 'react';
import { render } from '@testing-library/react';

import { ISceneNodeInternal, useStore, IAnimationComponentInternal } from '../../../../store';
import AnimationComponent, { toggleIsAnimationPaused } from '../AnimationComponent';
import { generateUUID } from '../../../../utils/mathUtils';

jest.mock('../../../../store', () => {
  const originalModule = jest.requireActual('../../../../store');
  return {
    ...originalModule,
  };
});

jest.mock('@react-three/fiber', () => {
  const originalModule = jest.requireActual('@react-three/fiber');
  return {
    ...originalModule,
    useLoader: jest.fn(),
    useFrame: jest.fn().mockImplementation((func) => {
      func();
    }),
  };
});

jest.useFakeTimers();

describe('toggleIsAnimationPaused', () => {
  test('tests if paused', () => {
    jest.mock('../../../../common/sceneComposerIdContext', () => ({
      useSceneComposerId: jest.fn(() => '3d74e3fe-9f8a-4190-a3c1-1b311a70ae30'),
    }));

    jest.mock('three', () => ({
      findByName: jest.fn(),
    }));
    const Mixer = new THREE.AnimationMixer({ animations: ['Action.013'] } as any);

    jest.mock('three', () => ({
      findByName: jest.fn(),
      clipAction: jest.fn(),
    }));

    expect(toggleIsAnimationPaused(Mixer, true, ['Action.013'])).toBe(undefined);
  });
});

const mockNode = {
  ref: 'ce7802dd-477e-4bae-b787-374aea3fce77',
  name: 'AIMIMATED_MIXER',
  transform: {
    position: [-0.2548017433916856, 0.1565180250896893, 1.9222755178165505],
    rotation: [0, 0, 0],
    scale: [1, 1, 1],
  },
  transformConstraint: {},
  properties: {},
  components: [
    {
      ref: 'f7bdb4b0-8627-4c83-825e-bace9874bc08',
      type: 'ModelRef',
      uri: 'AIMIMATED_MIXER.gltf',
      modelType: 'GLTF',
    },
    {
      selector: 1,
      ref: 'a2159735-991c-41da-a217-c39445a47b46',
      currentAnimations: ['Action.018'],
      uri: 'AIMIMATED_MIXER.gltf',
      type: 'Animation',
    },
  ],
  childRefs: [],
} as unknown as ISceneNodeInternal;

const object3D = new THREE.Object3D();
const sceneobject = new THREE.Object3D();

sceneobject.animations = { animations: 'Action.018' } as unknown as THREE.AnimationClip;
sceneobject.name = 'Scene';
object3D.add(sceneobject);
const baseState = {
  setScene: jest.fn().mockReturnValue(mockNode),
  getObject3DBySceneNodeRef: jest.fn().mockReturnValue(object3D),
  getObjectByName: jest.fn().mockReturnValue(new THREE.Scene()),
  document: {
    nodeMap: {
      'test-uuid': mockNode,
    },
  },
};

describe('<AnimationComponent />', () => {
  it('should test a simple animation component', () => {
    const mockComponent: IAnimationComponentInternal = {
      ref: generateUUID(),
      uri: 'Animated_Mixer.gltf',
      currentAnimations: ['Action.018'],
    } as IAnimationComponentInternal;

    useStore('default').setState(baseState);
    const { container } = render(<AnimationComponent node={mockNode} component={mockComponent as IComponent} />);

    expect(container).toMatchSnapshot();
  });
  it('should test more branches', () => {
    const mockComponent: IAnimationComponentInternal = {
      ref: generateUUID(),
      uri: 'Animated_Mixer.gltf',
      currentAnimations: ['Action.018'],
    } as IAnimationComponentInternal;
    jest.mock('three', () => ({
      findByName: jest.fn().mockReturnValue(''),
      clipAction: jest.fn().mockReturnValue(''),
    }));
    const mockObject3D = { s: 'action 1' };

    jest.mock('../../../../store', () => {
      const originalModule = jest.requireActual('../../../../store');

      const getObject3DMock = jest.fn();
      getObject3DMock.mockImplementation(() => {
        return mockObject3D;
      });
      return {
        ...originalModule,
        useEditorState: { getObject3DBySceneNodeRef: getObject3DMock },
      };
    });
    useStore('default').setState(baseState);
    const { container } = render(<AnimationComponent node={mockNode} component={mockComponent as IComponent} />);
    expect(container).toMatchSnapshot();
  });
});
