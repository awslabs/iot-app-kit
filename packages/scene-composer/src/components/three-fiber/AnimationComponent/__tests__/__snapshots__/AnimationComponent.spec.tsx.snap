// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnimationComponent /> should test a simple animation component 1`] = `<div />`;

exports[`<AnimationComponent /> should test more branches 1`] = `<div />`;
